records/2023-01-24.malicious-content-scanning.md

```diff
@@ -0,0 +1,61 @@
+# Malicious Content Scanning
+
+Date: 2023-01-24
+
+## Status
+
+**DRAFT**.
+
+_VALUES: (DRAFT, IN REVIEW, APPROVED, REJECTED)_
+
+## Context
+
+We are working on detecting malicious content in attached files, as described in [this issue card](https://app.zenhub.com/workspaces/notify-planning-614b3ad91bc2030015ed22f5/issues/gh/cds-snc/notification-planning/725). This is part of a larger effort to detect realtime misuse of Notify, which was a recommendation from a recent security audit.
+
+The SRE team has created the [scan-files](https://github.com/cds-snc/scan-files) project, which is a tool to help platform products scan files for malicious content. It includes terraform for [S3 scan object](https://github.com/cds-snc/scan-files/tree/main/module/s3-scan-object), which is sets up a lambda function that watches a given S3 bucket and scans each new file. The scan verdict is then written as a tag to the S3 object, and can be read when the S3 object is accessed.
+
+We have [tried getting the S3 scan object lambda function to run on our S3 bucket](https://github.com/cds-snc/notification-terraform/pull/600), but that did not work because we are using a different encryption key for each file in S3. This is because our S3 bucket for file uploads is public, and we need to prevent a user from being able to download and decrypt files they should not have access to.
+
+We are now calling the scan-files api directly from the Document Download API ([Draft PR](https://github.com/cds-snc/notification-document-download-api/pull/97)). This API can take a few seconds to return a response, and this time increases as file size increases. So we need to stop this from causing delaying the response when a user makes a requests to the Notify API.  
+
+We have explored using the `asyncio` library in python to create a new process to call the scan-files api [here](https://github.com/cds-snc/notification-document-download-api/pull/97/files#diff-a842b15bd72bea41437e64bda92d12262c8eca35af7e8adf111bd520fe7c3010R1). Then once the result is returned, we could write that result to the S3 object as a tag which can be read later. However, this approach did not work to reduce the response time for the Document Download API.
+
+It seems that it is [not possible to create background processes inside flask](https://flask.palletsprojects.com/en/2.2.x/async-await/#background-tasks).
+
+## Options
+
+Here are some of the options for scanning email attachments with the scan-files project.
+
+### Migrate Document Download API from Flask to Quart
```

Comment:
An option to migrate from Flask to Flask async could still be considered, as a complete migration to Quart might be more effort and Flask async might be enough to work for us.